gateway/reverse_proxy.go

```diff
@@ -1331,6 +1331,10 @@ func (p *ReverseProxy) WrappedServeHTTP(rw http.ResponseWriter, req *http.Reques
 	}
 
 	if createTransport {
+		if p.TykAPISpec.HTTPTransport != nil {
+			p.TykAPISpec.HTTPTransport.transport.CloseIdleConnections()
```

Comment:
That will close the idle connections but what about connections via `p.TykAPISpec.HTTPTransport.transport` still active serving requests? they will eventually become idle and remain.

Maybe there needs to be an extra wrapper above `transport` which tracks how many active requests there are, once it's zero, it should be safe to call `CloseIdleConnections` and know they're all closed. It'd probably have to run in some worker pool you could just push to via a channel... unless you can think of something simpler?

Reply:
Actually something simpler which should cover above case;

```golang
if p.TykAPISpec.HTTPTransport != nil {
  p.TykAPISpec.HTTPTransport.transport.DisableKeepAlives = true // this will stop active conns going to idle state			
  p.TykAPISpec.HTTPTransport.transport.CloseIdleConnections() // close idle conns created before keep alive was disabled
}
```

wdyt?

Reply:
Description of function says:
> CloseIdleConnections closes any connections which were previously connected from previous requests but are now sitting idle in a "keep-alive" state. It does not interrupt any connections currently in use.

So it will close connection in keep-alive state, which are not used. And it already knows about keep alive option.

HTTP transport, maintains pools of connections, and even when http client did the job, connection kept open by transport, so it can be re-used by other clients. This current leak caused by this "inactive" connections which kept alive just to be re-used, but since transport is not used anymore they wait forever. 

Also if there is "real" active connection from the user, it will not be closed by this function, but this is expected, since we do not want "disconnect" existing currently running requests.

Reply:
Ok take this scenario;

1. `CloseIdleConnections() ` closes all **current** idle connections, we now have 0 idle conns and 0 active conns
2. Before the new transport is assigned to `p.TykAPISpec.HTTPTransport ` there is already a request running which has passed `p.TykAPISpec.Unlock()` so is free to make a request on the old transport
3. Request completes and say `MaxIdleConnsPerHost` is one, this connection will go into idle state and leak.
4. By putting `DisableKeepAlives = true` ensures no connection will go into idle state after closing idle connections.

Can see in `net/http` it will stop connection from being queued into idle state

```golang
func (t *Transport) tryPutIdleConn(pconn *persistConn) error {
	if t.DisableKeepAlives || t.MaxIdleConnsPerHost < 0 {
		return errKeepAlivesDisabled
	}
```